### PR TITLE
Ch4120 1ncount

### PR DESCRIPTION
**Pull request by abusquets**

### CODE REVIEW

File: layerserver/api.py [hunk 1]
```diff
@@ -188,6 +190,23 @@ def transform_wkt(self, wkt, srid):
             geom.transform(trans)
         return geom
 
+    def _check_1n_count(self, qs):
+        if self.action == 'list':
+            virtual_fields = self.layer.virtual_fields.filter(
+                widget=DataBaseLayerVirtualField.WIDGET_CHOICES.relation1n)
+            for x in virtual_fields:
+                config = json.loads(x.widget_options)
+                self._virtual_fields.append(x.name)
```
Comment (manelclos) [inline layerserver/api.py]: @abusquets are we adding it to `_virtual_fields` even if `count is False`?

File: layerserver/serializers/databaselayer_content.py [hunk 2]
```diff
@@ -97,6 +97,14 @@ def append_token(self, url):
         return url
 
 
+class FixPropertiesSerializer(object):
```
Comment (manelclos) [inline layerserver/serializers/databaselayer_content.py]: @abusquets can you please explain why do we need this Serializer? also, if it is meant to be used as a Mixin, please add Mixin at the end of the class name.

File: layerserver/serializers/databaselayer_content.py [hunk 3]
```diff
@@ -138,13 +146,27 @@ def to_representation(self, obj):
...
-                if 'properties' in data and isinstance(data['properties'], dict):
-                    data['properties'][attribute] = self.fix_image_value(obj, attribute)
-                else:
-                    data[attribute] = self.fix_image_value(obj, attribute)
+                value = self.fix_image_value(obj, attribute)
+                self.append_value(data, attribute, value)
         return data
 
 
+class VFieldsSerializer(object):
```
Comment (manelclos) [inline layerserver/serializers/databaselayer_content.py]: @abusquets can we use the full name like in `VirtualFieldsSerializer`?

File: layerserver/serializers/databaselayer_content.py [hunk 4]
```diff
@@ -181,7 +203,12 @@ def apply_widgets(attrs, model, fields):
             attrs[field] = to_image_field(field, f)
 
 
-def create_dblayer_serializer(model, fields, id_field, read_only_fields):
+COMMON_MIXINS = (
```
Comment (manelclos) [inline layerserver/serializers/databaselayer_content.py]: @abusquets can we move this constant at the beginning of the file?

File: layerserver/api.py [hunk 5]
```diff
@@ -188,6 +190,23 @@ def transform_wkt(self, wkt, srid):
...
+                if 'count' in config and config['count'] is True:
+                    related_layer = DataBaseLayer.objects.filter(name=config['dblayer']).first()
+                    related_model = create_dblayer_model(related_layer)
+                    filter = {('%s' % config['dblayer_fk']): OuterRef(config['to_field'])}
+                    items = related_model.objects.filter(**filter).order_by().values(config['dblayer_fk'])
+                    count_items = items.annotate(c=Count(config['dblayer_fk'])).values('c')
+                    qs = qs.annotate(**{x.name: Coalesce(Subquery(count_items), Value(0))})
```
Comment (manelclos) [inline layerserver/api.py]: @abusquets can we replace these four lines with something like this?
```
qs = qs.annotate(**{'%s__count' % x.name: Count(related_model.related_query_name)})
```
Of course both models should be registered.

File: layerserver/api.py [hunk 6]
```diff
@@ -188,6 +190,23 @@ def transform_wkt(self, wkt, srid):
...
+                    print(qs.query)
```
[the remark this answers is not in the document]
Comment (abusquets) [reply layerserver/api.py]: ok

File: layerserver/api.py [hunk 7]
```diff
@@ -188,6 +190,23 @@ def transform_wkt(self, wkt, srid):
...
+    def _check_1n_count(self, qs):
+        if self.action == 'list':
+            virtual_fields = self.layer.virtual_fields.filter(
+                widget=DataBaseLayerVirtualField.WIDGET_CHOICES.relation1n)
+            for x in virtual_fields:
+                config = json.loads(x.widget_options)
+                self._virtual_fields.append(x.name)
+                if 'count' in config and config['count'] is True:
+                    related_layer = DataBaseLayer.objects.filter(name=config['dblayer']).first()
+                    related_model = create_dblayer_model(related_layer)
```
[the remark this answers is not in the document]
Comment (abusquets) [reply layerserver/api.py]: I think it is more important to verify the widget settings.
Here, if the widget is not set correctly, what should we do?

File: layerserver/api.py [hunk 8]
```diff
@@ -188,6 +190,23 @@ def transform_wkt(self, wkt, srid):
...
+                    filter = {('%s' % config['dblayer_fk']): OuterRef(config['to_field'])}
```
[the remark this answers is not in the document]
Comment (abusquets) [reply layerserver/api.py]: For me is ok. If is a problem for you I can change it

File: layerserver/api.py [hunk 9]
```diff
@@ -234,12 +254,14 @@ def get_pagination_class(self, layer):
                 return create_geojson_pagination_class(page_size=page_size, max_page_size=max_page_size)
 
     def get_serializer_class(self, *args, **kwargs):
+        # TODO: revisar method
```
[the remark this answers is not in the document]
Comment (abusquets) [reply layerserver/api.py]: Esta funcionalidad ya está implementada, debería eliminar TODO

File: layerserver/api.py (hunk 1 above)
Comment (abusquets) [reply layerserver/api.py]: I've checked and I don't like it. I'm going to change it

File: layerserver/serializers/databaselayer_content.py (hunk 3 above)
Comment (abusquets) [reply layerserver/serializers/databaselayer_content.py]: ok

File: layerserver/serializers/databaselayer_content.py (hunk 2 above)
Comment (abusquets) [reply layerserver/serializers/databaselayer_content.py]: Ok, I'm going to rename the class

File: layerserver/serializers/databaselayer_content.py (hunk 4 above)
Comment (abusquets) [reply layerserver/serializers/databaselayer_content.py]: ???
Another variables are in the middle of the script, and there is no problem,
ex. choices. why this case?
The classes aren't available at the top.

File: layerserver/serializers/databaselayer_content.py [hunk 10]
```diff
@@ -138,10 +146,18 @@ def to_representation(self, obj):
...
-                if 'properties' in data and isinstance(data['properties'], dict):
-                    data['properties'][attribute] = self.fix_image_value(obj, attribute)
-                else:
-                    data[attribute] = self.fix_image_value(obj, attribute)
+                value = self.fix_image_value(obj, attribute)
+                self.append_value(data, attribute, value)
+        return data
+
+
+class VirtualFieldSerializer(object):
```
Comment (manelclos) [inline layerserver/serializers/databaselayer_content.py]: @abusquets add Mixin at the end

**Pull request merged**